indexing-service/src/main/java/org/apache/druid/indexing/input/DruidSegmentInputFormat.java

```diff
@@ -55,14 +56,30 @@ public InputEntityReader createReader(
       File temporaryDirectory
   )
   {
-    return new DruidSegmentReader(
-        source,
-        indexIO,
-        inputRowSchema.getTimestampSpec(),
-        inputRowSchema.getDimensionsSpec(),
-        inputRowSchema.getColumnsFilter(),
-        dimFilter,
-        temporaryDirectory
+    // this method handles the case when the entity comes from a tombstone or from a regular segment
+    Preconditions.checkArgument(
+        source instanceof DruidSegmentInputEntity,
+        DruidSegmentInputEntity.class.getName() + " required, but "
+        + source.getClass().getName() + " provided."
     );
+
+    InputEntityReader retVal = null;
+    // source is tombstone case
+    if ((source instanceof DruidSegmentInputEntity) && ((DruidSegmentInputEntity) source).isFromTombstone()) {
+      retVal = new DruidTombstoneSegmentReader(source);
+    }
+    // source is not tombstone:
+    if (retVal == null) {
```

Comment:
Checking for `retVal == null` instead of just an `else` seems weird?

Reply:
yeah, weird indeed...made it an else